crates/ruff_linter/src/rules/flake8_pyi/rules/iter_method_return_iterable.rs

```diff
@@ -15,24 +15,46 @@ use crate::checkers::ast::Checker;
 /// `__iter__` methods should always should return an `Iterator` of some kind,
 /// not an `Iterable`.
 ///
-/// In Python, an `Iterator` is an object that has a `__next__` method, which
-/// provides a consistent interface for sequentially processing elements from
-/// a sequence or other iterable object. Meanwhile, an `Iterable` is an object
-/// with an `__iter__` method, which itself returns an `Iterator`.
+/// In Python, an `Iterable` is an object that has an `__iter__` method; an
+/// `Iterator` is an object that has `__iter__` and `__next__` methods. All
+/// `__iter__` methods are expected to return `Iterator`s in Python. Type
+/// checkers may not always recognize an object as being iterable if its
+/// `__iter__` method does not return an `Iterator`.
 ///
 /// Every `Iterator` is an `Iterable`, but not every `Iterable` is an `Iterator`.
-/// By returning an `Iterable` from `__iter__`, you may end up returning an
-/// object that doesn't implement `__next__`, which will cause a `TypeError`
-/// at runtime. For example, returning a `list` from `__iter__` will cause
-/// a `TypeError` when you call `__next__` on it, as a `list` is an `Iterable`,
-/// but not an `Iterator`.
+/// For example, `list` is an `Iterable`, but not an `Iterator`; you can obtain
+/// an iterator over a list's elements by calling `iter()` on the list:
+///
+/// ```pycon
+/// >>> import collections.abc
+/// >>> x = [42]
+/// >>> isinstance(x, collections.abc.Iterable)
+/// True
+/// >>> isinstance(x, collections.abc.Iterator)
+/// False
+/// >>> next(x)
+/// Traceback (most recent call last):
+///  File "<stdin>", line 1, in <module>
+/// TypeError: 'list' object is not an iterator
+/// >>> y = iter(x)
+/// >>> isinstance(y, collections.abc.Iterable)
+/// True
+/// >>> isinstance(y, collections.abc.Iterator)
+/// True
+/// >>> next(y)
+/// 42
+/// ```
+///
+/// Using `Iterable` rather than `Iterator` as a return type for an `__iter__`
+/// methods would imply that you would not necessarily be able to call `next()`
+/// on the returned object, violating the expectations of the interface.
```

Comment:
Nit: I'm okay with either.

```suggestion
/// on the returned object, violating the contract of the interface.
```

Reply:
I weakly prefer "expectations of the interface", so I'll stick with my current wording here 👍